client/src/main/java/io/pravega/client/security/auth/ValidJwtTokenHandlingStrategy.java

```diff
@@ -0,0 +1,142 @@
+        Exceptions.checkNotNullOrEmpty(token, "token");
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+
+        // Preconditions.checkNotNull(expTime, "No expiry is set in the delegation token");
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+    }
+
+    @Override
+    public String retrieveToken() {
+        if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry, so refreshing it");
+            return refreshToken();
+        } else {
+            return delegationToken.get().getValue();
+        }
+    }
+
+    @Override
+    public String refreshToken() {
+        resetToken(newToken());
+        return delegationToken.get().getValue();
+
+        /*if (delegationToken.get().getExpiryTime() != null) {
+            resetToken(newToken());
+        } else {
+            throw new IllegalStateException("Token expiry not set");
+        }
+        return delegationToken.get().getValue();*/
+    }
+
+    protected String newToken() {
+        return Futures.getAndHandleExceptions(
+                controllerClient.getOrRefreshDelegationTokenFor(scopeName, streamName), RuntimeException::new);
+    }
+
+    private synchronized boolean isTokenNearingExpiry() {
+        Long currentTokenExpirationTime = this.delegationToken.get().getExpiryTime();
+
+        // currentTokenExpirationTime can be null if the server returns a null delegation token
+        return currentTokenExpirationTime != null && isWithinRefreshThreshold(currentTokenExpirationTime);
+    }
+
+    private boolean isWithinRefreshThreshold(Long expirationTime) {
+        Preconditions.checkNotNull(expirationTime);
+        return isWithinRefreshThreshold(Instant.now(), Instant.ofEpochSecond(expirationTime));
+    }
+
+    @VisibleForTesting
+    boolean isWithinRefreshThreshold(Instant currentInstant, Instant expiration) {
+        return currentInstant.plusSeconds(REFRESH_THRESHOLD).getEpochSecond() >= expiration.getEpochSecond();
+    }
+
+    protected void resetToken(String newToken) {
+        delegationToken.set(new DelegationToken(newToken, extractExpirationTime(newToken)));
+    }
+
+    @VisibleForTesting
+    Long extractExpirationTime(@NonNull String token) {
+        if (token == null || token.trim().equals("")) {
+            return null;
+        }
+        String[] tokenParts = token.split("\\.");
+
+        // JWT is of the format abc.def.ghe. The middle part is the body.
+        String encodedBody = tokenParts[1];
+        String decodedBody = new String(Base64.getDecoder().decode(encodedBody));
+
+        JwtBody jwt = new Gson().fromJson(decodedBody, JwtBody.class);
```

Comment:
The `JwtTokenHandlingStrategy` localizes the JWT token handling strategy in its own class. If we use a different delegation token format later, we can add a new strategy implementation class and activate it in the `DelegationTokenProviderImpl` class. Hope that addresses your concern about the format lock-in. 

As for the use of the transitive Gson dependency, I've now added Gson as a dependency in the client. We parse the JWT JSON-formatted body to obtain the token expiration time, in order to find out if the token is nearing expiry; To parse the JSON, we are using Gson deserialization, and therefore need it as a dependency. Hope that is OK.

Reply:
There isn't any configuration parameter to be specified by the client application unless the client application developer intends to modify the time-based threshold for a refreshing token (refreshes token if it is about to expire in 5 seconds, by default). 

Plus the format used for delegation tokens - JWT currently - is opaque to the client application. The application doesn't care or need to know. It's an internal thing between the client library code and the Controller. 

A good thing about parsing the expiry time from the token alone is that we can potentially use/pass other params through the delegation token, without having to change the Controller gRPC API that returns delegation token for the scope and stream. For example, we might decide later to also verify the entity that has created the delegation token in the future. Changing the Controller API for each of such changes may not work out very well. 

Please consider these aspects as well, and reassess the mechanism. Please let me know.

Reply:
> There isn't any configuration parameter to be specified by the client application unless the client application developer intends to modify the time-based threshold for a refreshing token (refreshes token if it is about to expire in 5 seconds, by default).

Correct, there is no reason to modify that.

> Plus the format used for delegation tokens - JWT currently - is opaque to the client application. The application doesn't care or need to know. It's an internal thing between the client library code and the Controller.

Isn't this change undoing all of that? The client library is now using json to decode the jwt token. Instead of doing that, the controller could simply return two fields in the reply.

> A good thing about parsing the expiry time from the token alone is that we can potentially use/pass other params through the delegation token, without having to change the Controller gRPC API

I agree we should be able to use/pass other parameters through the delegation token to the server. I don't agree that we should use them to pass data to the client. The client-controller API is already versioned, supports adding fields, and has a defined protocol. All other changes in client-server communication happen there. There is no advantage in circumenting that for one code path.

If on the other hand the token remains opaque to the client, then we can put whatever we want in there without the client code knowing or caring about it. The server can then use this additional information as appropriate.

Reply:
I see the value of making the token opaque to the client. 

However, the expiry time we are extracting is part of the JWT formatted token, so I believe it'd be overkill to also send it out of band via the Controller's response to the client's request for delegation token, when it is already part of the token. The JWT token itself is a self-contained object containing everything necessary to process the token, including the registered claim `exp` (expiration time) (https://tools.ietf.org/html/rfc7519#page-9). 

Also, to clarify about my assertion about passing other data via delegation token, I was only referring to data necessary for processing a delegation token, such as registered claims listed [here](https://tools.ietf.org/html/rfc7519#page-9). 

That being said, I've modified the code now to avoid the dependency for `Gson`. 

I hope that works.